iml-services/iml-devices/src/change.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2019 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    ops::Deref,
+};
+
+#[derive(Debug, PartialEq)]
+pub enum Change<T> {
+    Add(T),
+    Update(T),
+    Remove(T),
+}
+
+impl<T> Deref for Change<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        match self {
+            Change::Add(x) | Change::Update(x) | Change::Remove(x) => x,
+        }
+    }
+}
+
+impl<T> Change<T> {
+    pub fn map<R, F: Fn(&T) -> R>(&self, f: F) -> Change<R> {
+        match self {
+            Change::Add(x) => Change::Add(f(x)),
+            Change::Update(x) => Change::Update(f(x)),
+            Change::Remove(x) => Change::Remove(f(x)),
+        }
+    }
+    pub fn is_add(&self) -> bool {
+        match self {
+            Change::Add(_) => true,
+            _ => false,
+        }
+    }
+}
+
+pub fn get_changes_values<
+    'a,
+    'b,
+    K: std::cmp::Eq + std::cmp::Ord + std::fmt::Debug,
+    V: std::cmp::Eq + std::fmt::Debug,
+>(
+    old: &'b BTreeMap<&'a K, V>,
+    new: &'b BTreeMap<&'a K, V>,
+) -> Vec<Change<&'b V>> {
+    let old_ids: BTreeSet<_> = old.keys().collect();
+    let new_ids: BTreeSet<_> = new.keys().collect();
+
+    let to_change = new_ids
+        .intersection(&old_ids)
+        .filter(|&&k| old.get(k) != new.get(k))
+        .inspect(|&&k| tracing::debug!("not equal. old: {:?}, new {:?}", old.get(k), new.get(k)))
+        .map(|&k| Change::Update(new.get(k).unwrap()));
+
+    tracing::trace!("old ids: {:?}. new ids: {:?}", old_ids, new_ids);
+
+    let to_remove = old_ids
+        .difference(&new_ids)
+        .map(|&k| Change::Remove(old.get(k).unwrap()));
+
+    let to_add = new_ids
+        .difference(&old_ids)
+        .map(|&k| Change::Add(new.get(k).unwrap()));
+
+    to_change.chain(to_remove).chain(to_add).collect()
+}
+pub fn get_changes<
+    'a,
+    'b,
+    K: std::cmp::Eq + std::cmp::Ord + Clone,
+    V: std::cmp::Eq + std::fmt::Debug,
+>(
+    old: &'b BTreeMap<&'a K, V>,
+    new: &'b BTreeMap<&'a K, V>,
+) -> Vec<Change<K>> {
+    let old_ids: BTreeSet<_> = old.keys().collect();
+    let new_ids: BTreeSet<_> = new.keys().collect();
+
+    let to_change = new_ids
+        .intersection(&old_ids)
+        .filter(|&&k| old.get(k) != new.get(k))
+        .inspect(|&&k| tracing::debug!("not equal. old: {:?}, new {:?}", old.get(k), new.get(k)))
+        .map(|&&k| Change::Update(k.clone()));
+
+    let to_remove = old_ids
+        .difference(&new_ids)
+        .map(|&&k| Change::Remove(k.clone()));
+
+    let to_add = new_ids
+        .difference(&old_ids)
+        .map(|&&k| Change::Add(k.clone()));
+
+    to_change.chain(to_remove).chain(to_add).collect()
+}
+
+#[cfg(test)]
```

Comment:
There is some redundancy in the tests. There are tests for the same functionality, but with different types: https://github.com/whamcloud/integrated-manager-for-lustre/pull/1176/files#diff-9b7dc341343be7c2d3abfaf18897c3c6R474

Reply:
I can't see what you're linking to, it just shows the entire diff.

Reply:
![thetest](https://user-images.githubusercontent.com/131844/75134409-e6a10300-56e6-11ea-8c7a-17fadec5082b.png)

Reply:
While making sure it works I've wanted to test all levels. In addition to these tests, in `db.rs` we have `deser_fixture` and unknown data in JSON. And it was in practice a source of logical mistakes too - where I've made fixtures that weren't correct. More tests makes it easier to find such mistakes.

Reply:
Well, there are no levels, it is a generic function what is being tested. Whatever type you substitute you're basically testing the same thing. More tests are useful when they are testing different things or corner cases, but here they are doing the same. It would be useful to have few tests with different input data (not just different types, but different sets). 

I am not against _these_ simple tests here, I would keep them rather than those in `db.rs` :)
 It is just strange to have similar tests for the same code in different places.